Type Predicate Expressions iteration 3 (#669)
neo4j/docs-cypher: modules/ROOT/pages/syntax/expressions.adoc

1+d|Rows: 1
|===

> Note that closed dynamic types (`INNER_TYPE_1 | INNER_TYPE_2...`) cannot be appended generically with `NOT NULL`, instead all inner types must be nullable, or all not nullable. A syntax error will be raised if the inner types are not of the same nullability.

Reviewer: How about adding a small why this is not allowed, e.g. "since null values could not be distinguished between the types"

Reviewer: I think this sentence is very dense.
```suggestion
Note that closed dynamic types (`INNER_TYPE_1 | INNER_TYPE_2...`) cannot be declared as NOT NULL. Instead, the inner types can be declared as not nullable: [add example]

Note that all inner types must be nullable, or all not nullable, since null values can't  be attributed to a specific type. A syntax error will be raised if the inner types are not of the same nullability.
```

1+d|Rows: 1
|===

Closed dynamic union types (`INNER_TYPE_1 | INNER_TYPE_2...`) cannot be declared as `NOT NULL`. Instead, all the inner types should be individually declared as not nullable to achieve this behavior.

Note that all inner types in a closed dynamic union must be either nullable, or not nullable. This is because `null` values cannot be attributed to a specific type. A syntax error will be raised if the inner types are not of the same nullability.

[source, cypher, role=test-fail]
----
RETURN 1 IS :: INTEGER NOT NULL | FLOAT
----

.Error message

> [source, output, role="noheader"]

Reviewer 2: ```suggestion
[source, output]
```
Do you want a header, or do you not want it? XD

At all rates, if you don't want one, you just need to omit the title for the block (line 1273). `role=noheader` does nothing nowadays.

----

> All types in a Closed Dynamic Union must be nullable, or be appended with `NOT NULL`

Reviewer 2: ```suggestion
All types in a Closed Dynamic Union must be nullable, or be appended with `NOT NULL`.
```
Worth double checking if the error is also lacking a full stop in the source.

2+d|Rows: 1
|===

[[type-predicate-expressions-closed-dynamic-unions]]
=== Closed Dynamic Unions

_This feature was introduced in Neo4j 5.11._

> Closed dynamic union types allow for the testing of multiple types in the same predicate.

Reviewer 2: ```suggestion
With closed dynamic union types you can test for multiple types in the same predicate.
```

Author: This change feels wrong when read allowed

> This allows for the checking of mixed types:

Author: Same as above, doesn't feel natural in English

> This allows for the checking of mixed types:

Reviewer 2: I'd just drop this sentence and replace it with a `For example:`. Either in its own paragraph or together with the opening sentence.

3+d|Rows: 1
|===

Lists can be combined with closed dynamic union types to create tests for heterogeneous lists.

[source, cypher]
----
WITH [1, 0, true, false] AS mixedList
RETURN booleanList IS :: LIST<BOOLEAN | INTEGER> as isMixedList

> ----

Reviewer 2: There's also a failing test regarding this example. 
One of `mixedList` and `booleanList` need to turn into the other :)
